seqfold/fold.go

```diff
@@ -0,0 +1,875 @@
+package seqfold
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"strings"
+)
+
+// Fold the DNA sequence and return the lowest free energy score.
+//
+// Based on the approach described in:
+// Zuker and Stiegler, 1981
+// https://www.ncbi.nlm.nih.gov/pmc/articles/PMC326673/pdf/nar00394-0137.pdf
+//
+// If the sequence is 50 or more bp long, "isolated" matching bp
+// are ignored in V(i,j). This is based on an approach described in:
+// Mathews, Sabina, Zuker and Turner, 1999
+// https://www.ncbi.nlm.nih.gov/pubmed/10329189
+//
+// Args:
+//     seq: The sequence to Fold
+// Keyword Args:
+//     temp: The temperature the Fold takes place in, in Celsius
+// Returns:
+//     List[Struct]: A list of structures. Stacks, bulges, hairpins, etc.
+func Fold(seq string, temp float64) []NucleicAcidStructure {
+	//v_cache, w_cache := cache(seq, temp)
+	//n := len(seq)
+	fc, err := NewFoldingContext(seq, temp)
+	if err != nil {
+		fmt.Printf("%+v\n", err)
+		os.Exit(1)
```

Comment:
should return error instead of `os.Exit(1)`

Reply:
Ah yes, I was still thinking in terms of command line tools as the python code.

There are also a bunch of panics I left in place of python `assert`.
I am not entirely sure how to deal with those, it is probably an error we cannot recover from so either we bubble it up to the caller or recover at some point.

I would prefer the error all the way but it's probably a big refactor.